pyro/contrib/forecast/evaluate.py

```diff
@@ -0,0 +1,163 @@
+             forecaster_options={}):
+    """
+    Backtest a forecasting model on a moving window of (train,test) data.
+
+    :param data: A tensor dataset with time dimension -2.
+    :type data: ~torch.Tensor
+    :param covariates: A tensor of covariates with time dimension -2.
+        For models not using covariates, pass a shaped empty tensor
+        ``torch.empty(duration, 0)``.
+    :type covariates: ~torch.Tensor
+    :param model:
+    :type model: ~pyro.contrib.forecast.forecaster.ForecastingModel
+
+    :param dict metrics: A dictionary mapping metric name to metric function.
+        The metric function should input a forecast ``pred`` and ground
+        ``truth`` and can output anything, often a number. Example metrics
+        include: :func:`eval_mae`, :func:`eval_rmse`, and :func:`eval_crps`.
+    :param callable transform: An optional transform to apply before computing
+        metrics. If provided this will be applied as
+        ``pred, truth = transform(pred, truth)``.
+    :param int train_window: Size of the training window. Be default trains
+        from beginning of data.
+    :param int min_train_window: If ``train_window`` is None, this specifies
+        the min training window size. Defaults to 1.
+    :param int test_window: Size of the test window. By default forecasts to
+        end of data.
+    :param int min_test_window: If ``test_window`` is None, this specifies
+        the min test window size. Defaults to 1.
+    :param int stride: Optional stride for test/train split. Defaults to 1.
+    :param int seed: Random number seed.
+    :param int num_samples: Number of samples for forecast.
+    :param dict forecaster_options: Options to pass to forecaster. See
+        :class:`~pyro.contrib.forecaster.Forecaster` for details.
+
+    :returns: A dict mapping (train_begin, split, test_end) triples to
+        dictionaries of evaluation data. Caller is responsible for aggregating
+        the per-window metrics.
+    :rtype: dict
+    """
+    assert data.size(-2) == covariates.size(-2)
+    assert isinstance(min_train_window, int) and min_train_window >= 1
+    assert isinstance(min_test_window, int) and min_test_window >= 1
+    if metrics is None:
+        metrics = DEFAULT_METRICS
+    assert metrics, "no metrics specified"
+
+    duration = data.size(-2)
+    if test_window is None:
+        stop = duration - min_test_window
+    else:
+        stop = duration - test_window
+    if train_window is None:
+        start = min_train_window
+    else:
+        start = train_window
+
+    result = {}
+    for t1 in range(start, stop, stride):
+        t0 = 0 if train_window is None else t1 - train_window
+        t2 = duration if test_window is None else t1 + test_window
+        assert 0 <= t0 < t1 < t2 <= duration
+        logger.info("Evaluating on window ({}, {}, {})".format(t0, t1, t2))
+
+        # Train a forecaster on the training window.
+        pyro.set_rng_seed(seed)
```

Comment:
Is it okay to reuse the seed for each window? We can move it outside but I guess the `seed` argument below won't be particularly meaningful in that case.

Reply:
I think it is safer to reset the seed for each window, that way we can e.g. compute with large stride, then fill in the dict with smaller stride later. You're right that "seed" is then duplicated in all records. I'll switch to adding seed as an index column, that way we can test with multiple seeds and eventually store in a single dict.

Reply:
OK, I've switched to a list output format. That way we can run with multiple random seeds and aggregate multiple experiments.